## Table of Contents:
#### 01. Importing the libraries
#### 02. Importing the dataframe
#### 03. Aggregating the 'order_number' column
#### 04. Creating a loyalty flag for the existing customers
#### 05. Finding the average product price of each loyalty category
#### 06. Creating a spending flag for each user
#### 07. Creating an order-frequency flag for each user
#### 08. Exporting the dataframe

## 01. Importing the libraries

In [1]:
# importing the libraries

import pandas as pd
import numpy as np
import os

## 02. Importing the dataframe

In [2]:
# defining the path

path=r'/Users/sanju/Documents/Jul 2023 Instacart Basket Analysis/02 Data'

In [3]:
# importing the 'orders_products_merged_derived.pkl' dataframe

ords_prods_merge=pd.read_pickle(os.path.join(path,'Prepared Data','orders_products_merged_derived.pkl'))

In [4]:
# checking the imported dataframe

ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly Busy  Regularly Busy        Average orders  
1  Regularly Busy      Least Busy        Average orders  
2  Regularly Busy      Least Busy           Most orders  
3      Least Busy      Least Busy        Average orders  
4      Least Busy      Least Busy           Most orders

In [6]:
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 18 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int32   
 1   user_id                 int32   
 2   order_number            int8    
 3   orders_day_of_week      int8    
 4   order_hour_of_day       int8    
 5   days_since_prior_order  float16 
 6   product_id              int32   
 7   add_to_cart_order       int32   
 8   reordered               int8    
 9   _merge                  category
 10  product_name            object  
 11  aisle_id                int16   
 12  department_id           int16   
 13  prices                  float32 
 14  price_range_loc         object  
 15  busiest_day             object  
 16  busiest_days            object  
 17  busiest_period_of_day   object  
dtypes: category(1), float16(1), float32(1), int16(2), int32(4), int8(4), object(5)
memory usage: 2.4+ GB


## 03. Aggregating the 'order_number' column

In [7]:
# aggregating the 'order_number column' by Mean and grouped by 'department_id'

ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

#### The minimum order number of the subset was 11.29 for the department_id 17 (household), whereas, for the entire dataframe it's 15.2 for the department_id 5 (alcohol).
#### Similarly, the maximum order number of the subset was 19.46 for the department_id 16 (dairy eggs), whereas, for the entire dataframe it's 22.90 for the department_id 21 whose category is missing. This shows that there are quite many products with no specified categories.

## 04. Creating a loyalty flag for the existing customers

In [8]:
# grouping data by 'user_id' and creating 'max_order' column for each user_id

ords_prods_merge['max_order']=ords_prods_merge.groupby('user_id')['order_number'].transform(np.max)

In [9]:
# checking the result

ords_prods_merge.head(30)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                      NaN         196                  1          0   both   
1                     15.0         196                  1          1   both   
2                     21.0         196                  1          1   both   
3                     29.0         196                  1          1   both   
4                     28.0         196                  1          1   both   
5                     19.0         196                  1          1   both   
6                     20.0         196                  1          1   both   
7                     14.0         196                  2          1   both   
8                      0.0         196                  4          1   both   
9                     30.0         196                  1          1   both   
10                     7.0         196                  2          0   both   
11                     8.0         196                  6          1   both   
12                     7.0         196                  1          1   both   
13                     7.0         196                  1          1   both   
14                    14.0         196                  1          1   both   
15                     6.0         196                  3          0   both   
16                     8.0         196                 15          1   both   
17                     6.0         196                  1          1   both   
18                     6.0         196                  2          0   both   
19   

In [10]:
# creating a flag with the loc() function

ords_prods_merge.loc[ords_prods_merge['max_order']>40,'loyalty_flag']='Loyal Customer'

In [11]:
ords_prods_merge.loc[(ords_prods_merge['max_order']<=40) & (ords_prods_merge['max_order']>10),'loyalty_flag']='Regular Customer'

In [12]:
ords_prods_merge.loc[ords_prods_merge['max_order']<=10,'loyalty_flag']='New Customer'

In [13]:
# checking the frequency of the 'loyalty_flag' column

ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

Regular Customer    15876776
Loyal Customer      10284093
New Customer         6243990
Name: loyalty_flag, dtype: int64

In [14]:
ords_prods_merge[['user_id','loyalty_flag','order_number']].head(30)

user_id      loyalty_flag  order_number
0         1      New Customer             1
1         1      New Customer             2
2         1      New Customer             3
3         1      New Customer             4
4         1      New Customer             5
5         1      New Customer             6
6         1      New Customer             7
7         1      New Customer             8
8         1      New Customer             9
9         1      New Customer            10
10       15  Regular Customer            15
11       15  Regular Customer            17
12       15  Regular Customer            18
13       15  Regular Customer            21
14       15  Regular Customer            22
15       19      New Customer             2
16       19      New Customer             5
17       19      New Customer             7
18       21  Regular Customer            10
19       31  Regular Customer            10
20       31  Regular Customer            17
21       43  Regular Customer             6
22       43  Regular Customer             9
23       52  Regular Customer             1
24       52  Regular Customer             2
25       52  Regular Customer             3
26       52  Regular Customer             5
27       52  Regular Customer             7
28       52  Regular Customer             9
29       52  Regular Customer            10

## 05. Finding the average product price of each loyalty category

In [16]:
# grouping the data by 'loyalty_flag' and finding the Mean of the 'price' column

ords_prods_merge.groupby('loyalty_flag')['prices'].mean()

loyalty_flag
Loyal Customer      10.386335
New Customer        13.294669
Regular Customer    12.495717
Name: prices, dtype: float32

## 06. Creating a spending flag for each user

In [17]:
# grouping the data by 'user_id' and creating a new 'avg_price' column for each user

ords_prods_merge['avg_price']=ords_prods_merge.groupby('user_id')['prices'].transform(np.mean)

In [18]:
# checking the output

ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                      NaN         196                  1          0   both   
1                     15.0         196                  1          1   both   
2                     21.0         196                  1          1   both   
3                     29.0         196                  1          1   both   
4                     28.0         196                  1          1   both   
5                     19.0         196                  1          1   both   
6                     20.0         196                  1          1   both   
7                     14.0         196                  2          1   both   
8                      0.0         196                  4          1   both   
9                     30.0         196                  1          1   both   
10                     7.0         196                  2          0   both   
11                     8.0         196                  6          1   both   
12                     7.0         196                  1          1   both   
13                     7.0         196                  1          1   both   
14                    14.0         196                  1          1   both   
15                     6.0         196                  3          0   both   
16                     8.0         196                 15          1   both   
17                     6.0         196                  1          1   both   
18                     6.0         196                  2          0   both   
19                     7.0         196                 17          0   both   

    ... aisle_id  department_id  prices    price_range_loc     busiest_day  \
0   ...       77              7     9.0  Mid-range product  Regularly Busy   
1   ...       77              7     9.0  Mid-range product  Regularly Busy   
2   ...       77              7     9.0  Mid-range product  Regularly Busy   
3   ...       77              7     9.0  Mid-range product      Least Busy   
4   ...       77              7     9.0  Mid-range product      Least Busy   
5   ...       77              7     9.0  Mid-range product  Regularly Busy   
6   ...       77              7     9.0  Mid-range product  Regularly Busy   
7   ...       77              7     9.0  Mid-range product  Regularly Busy   
8  

In [19]:
# creating a flag with the loc() function

ords_prods_merge.loc[ords_prods_merge['avg_price']<10,'spending_flag']='Low spender'

In [20]:
ords_prods_merge.loc[ords_prods_merge['avg_price']>=10,'spending_flag']='High spender'

In [21]:
# checking the frequency of the 'spending_flag' column

ords_prods_merge['spending_flag'].value_counts(dropna=False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

In [22]:
# checking the output

ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                      NaN         196                  1          0   both   
1                     15.0         196                  1          1   both   
2                     21.0         196                  1          1   both   
3                     29.0         196                  1          1   both   
4                     28.0         196                  1          1   both   
5                     19.0         196                  1          1   both   
6                     20.0         196                  1          1   both   
7                     14.0         196                  2          1   both   
8                      0.0         196                  4          1   both   
9                     30.0         196                  1          1   both   
10                     7.0         196                  2          0   both   
11                     8.0         196                  6          1   both   
12                     7.0         196                  1          1   both   
13                     7.0         196                  1          1   both   
14                    14.0         196                  1          1   both   
15                     6.0         196                  3          0   both   
16                     8.0         196                 15          1   both   
17                     6.0         196                  1          1   both   
18                     6.0         196                  2          0   both   
19                     7.0         196                 17          0   both   

    ... department_id  prices    price_range_loc     busiest_day  \
0   ...             7     9.0  Mid-range product  Regularly Busy   
1   ...             7     9.0  Mid-range product  Regularly Busy   
2   ...             7     9.0  Mid-range product  Regularly Busy   
3   ...             7     9.0  Mid-range product      Least Busy   
4   ...             7     9.0  Mid-range product      Least Busy   
5   ...             7     9.0  Mid-range product  Regularly Busy   
6   ...             7     9.0  Mid-range product  Regularly Busy   
7   ...             7     9.0  Mid-range product  Regularly Busy   
8   ...             7     9.0  Mid-range product  Regularly Busy   
9   ...             7    

## 07. Creating an order-frequency flag for each user

In [23]:
# grouping the data by 'user_id' and creating a new 'avg_price' column for each user

ords_prods_merge['median_order_days']=ords_prods_merge.groupby('user_id')['days_since_prior_order'].transform(np.median)

In [24]:
# checking the output

ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                      NaN         196                  1          0   both   
1                     15.0         196                  1          1   both   
2                     21.0         196                  1          1   both   
3                     29.0         196                  1          1   both   
4                     28.0         196                  1          1   both   
5                     19.0         196                  1          1   both   
6                     20.0         196                  1          1   both   
7                     14.0         196                  2          1   both   
8                      0.0         196                  4          1   both   
9                     30.0         196                  1          1   both   
10                     7.0         196                  2          0   both   
11                     8.0         196                  6          1   both   
12                     7.0         196                  1          1   both   
13                     7.0         196                  1          1   both   
14                    14.0         196                  1          1   both   
15                     6.0         196                  3          0   both   
16                     8.0         196                 15          1   both   
17                     6.0         196                  1          1   both   
18                     6.0         196                  2          0   both   
19                     7.0         196                 17          0   both   

    ... prices    price_range_loc     busiest_day    busiest_days  \
0   ...    9.0  Mid-range product  Regularly Busy  Regularly Busy   
1   ...    9.0  Mid-range product  Regularly Busy      Least Busy   
2   ...    9.0  Mid-range product  Regularly Busy      Least Busy   
3   ...    9.0  Mid-range product      Least Busy      Least Busy   
4   ...    9.0  Mid-range product      Least Busy      Least Busy   
5   ...    9.0  Mid-range product  Regularly Busy  Regularly Busy   
6   ...    9.0  Mid-range product  Regularly Busy     Busiest Day   
7   ...    9.0  Mid-range product  Regularly Busy     Busiest Day   
8   ...    9.0  Mid-range product  Regularly Busy     Busiest Day   
9   ...    9.0 

In [25]:
# creating a flag with the loc() function

ords_prods_merge.loc[ords_prods_merge['median_order_days']>20, 'order_frequency_flag']='Non-frequent customer'

In [27]:
ords_prods_merge.loc[(ords_prods_merge['median_order_days']<=20) & (ords_prods_merge['median_order_days']>10), 'order_frequency_flag']='Regular customer'

In [28]:
ords_prods_merge.loc[ords_prods_merge['median_order_days']<=10,'order_frequency_flag']='Frequent customer'

In [30]:
# checking the frequency of the 'order_frequency_flag' column

ords_prods_merge['order_frequency_flag'].value_counts(dropna=False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_frequency_flag, dtype: int64

In [31]:
# checking the null values

print(ords_prods_merge[ords_prods_merge.order_frequency_flag.isnull()])

          order_id  user_id  order_number  orders_day_of_week  \
13645692    895835   159838             1                   0   
17251990    895835   159838             1                   0   
17622767    895835   159838             1                   0   
24138593    895835   159838             1                   0   
25880002    895835   159838             1                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
13645692                 17                     NaN       10749   
17251990                 17                     NaN       33401   
17622767                 17                     NaN       23695   
24138593                 17                     NaN       21334   
25880002                 17                     NaN       22198   

          add_to_cart_order  reordered _merge  ...    price_range_loc  \
13645692                  3          0   both  ...  Mid-range product   
17251990                  6          0   both  ...  Mid-rang

#### We can see above that it's the first time order for all the 5 customers. Hence it is safe to categorize these customers as non-frequent customers.

In [32]:
# changing the NaN values from the 'order_frequency_flag'

ords_prods_merge['order_frequency_flag'].fillna("Non-frequent customer", inplace = True)

In [33]:
# checking the output

ords_prods_merge['order_frequency_flag'].value_counts(dropna=False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636442
Name: order_frequency_flag, dtype: int64

## 08. Exporting the dataframe

In [35]:
# exporting the dataframe as a pickle file

ords_prods_merge.to_pickle(os.path.join(path,'Prepared Data','orders_products_aggregated.pkl'))